bluesky/callbacks/mpl_plotting.py

```diff
@@ -124,9 +126,12 @@ def event(self, doc):
         if self.x == 'time' and self._epoch == 'run':
             new_x -= self._epoch_offset
 
+        self.update(new_x, new_y)
+        super().event(doc)
+
+    def update(self, new_x, new_y):
```

Comment:
Note that this change is so that we have a somewhat uniform interface, `update` will take in the raw data types and do the expected thing (add them to a list, plot them, etc). This allows us to use these callbacks without the event model, eg if you had already unpacked the data somewhere.

Reply:
The input to `update` below is of an entirely different type than the `update` here.  I do not fully understand this use case.  You still need `Start` and `Descriptor` documents to set up the `Callback` objects to be ready to take in data.  In what case would you have `Start` and `Descriptor` documents but not `Event` documents?  

In general callbacks may configure them selves based on the `Start` and `Descriptor` documents or do very different things based on which stream the `Event` is in there may not be a single `update` method with a fixed signature (see BestEffortCallback for example).

Reply:
My hope was that a) we could use callbacks outside the event model by calling `update` with the actual data. It would be upon the user/dev to properly parse the above `start` and `descriptor` code and make certain that all the state was created as needed. This would mean that we could use this on data which didn't adhere to the event model. I'd really prefer not to have to rewrite everything just so I can pass raw data to it.

My understanding of `BestEffortCallback` is that it is a bit of an outlier since it is a gestalt of callbacks.  I'd imagine that most of these base mpl driven classes will not do vastly different things based off of the `Start` and `Descriptor` since they are designed to produce a specific plot from a specific type of data.

Reply:
If that is the case then you should write functions independent of the callbacks that work on raw data and then use those functions in the callback.

The start/descriptor (plus maybe some values passed at init time) should be providing instructions to the callback about how to unpack the events and the `event` method is already the 'update' method.  I do not see the advantage of standardizing on a second method that requires understanding the internal state of the callback to use.  One of the biggest strengths of the current callback design is that users of the callbacks do not need to know _anything_ about what goes on inside (in principle).  We should be moving to making those abstractions less leaky.

BEC is probably prototypical of callbacks going forward with more brains / dispatch logic rolled in.

Reply:
I don't think `CallbackBase` should be used "outside of the event model." Its role its to communicate between the event model and generic Python functions. Anything outside of the event model could use some simpler callback mechanism without the four methods named for event model documents, right?

Reply:
I guess, are there plans to break the callbacks so as to separate the update and event model logic?

Reply:
Not really. The `update` method was added in a refactor intended to make it easier for subclasses to do simple arithmetic on the data before plotting. I've recently thought it would be better to do that through a callback hook (`prepare_data` or something). Like `LivePlot` in general, this API is not particularly well thought out and not intended to be an extensible framework, more like a simple example. The only rock-solid API is `RunEngine.subscribe` and, equivalently, `CallbackBase`.

Reply:
I'm thinking about this specifically since there are other applications of streaming which could benefit from the mpl wrapper chunks that this has done. (eg I really wouldn't want to try to re-implement the image viewer stuff or the scatter plot if I didn't have to and a solution to one seems like a solution to both)

Reply:
I see. What would it look like to make an image viewer (with some `update` method) that knows nothing of the event model and then an adapter between said image viewer and the event model which inherits from `CallbackBase` and calls `CJsAwesomeImageViewer.update`?

Reply:
We really need to re-suscitate xray-vision...

Reply:
That works too!
For live plot we could move most of the `start` document stuff to the init
```python
self.x_data, self.y_data = [], []
label = " :: ".join(
            [name for name in self.legend_keys])
kwargs = ChainMap(self.kwargs, {'label': label})
self.current_line, = self.ax.plot([], [], **kwargs)
self.lines.append(self.current_line)
self.legend = self.ax.legend(
            loc=0, title=self.legend_title).draggable()
```
Then have the `start` method overwrite the label/remake the legend. Have the event method unpack the data and hand it to update.

Reply:
Note that this may also help with `LiveImage` as well, since we can create a version which assumes that the data is already loaded, and have a broker driven class which loads/feeds data to the `LiveImage` class.

Reply:
> Note that this may also help with LiveImage as well, since we can create a version which assumes that the data is already loaded, and have a broker driven class which loads/feeds data to the LiveImage class.

I don't understand.  Broker _also_ emits documents so it should be the same code for live or post-fact.

> Then have the start method overwrite the label/remake the legend. Have the event method unpack the data and hand it to update.

To be clear 'update' is not part of the callback API

> For live plot we could move most of the start document stuff to the init

Why?  You do not have access to the information in the start document at `__init__` time

Reply:
For `LiveImage` we have a broker filling events, if we took only data then we wouldn't need to worry about where the data was/which broker to hook into.

I understand it isn't a part of the callback API. My point is that there could be two APIs:
1. Classes which wrap MPL plots and take streaming data via an `update` method. Information not handled in the streaming data can be handled at init and changed later (via attributes). Note that this API doesn't really care where the data is coming from (eg some raw streaming firehose somewhere).
1. Classes which take event model and turn it into data (potentially by using a broker to fill events) which is plumbed into the first API's `update` and metadata which is used to change the first API's state to have the desired info (change the title, legends, etc.).

In the non event model case the `__init__` time info could be entered manually (eg, I want the title to be 'x' because that is the name of the stream to which I'm attached). In the case of the event model, this could be defaulted sanely, especially since it will be quickly overridden by the info from the documents.

tl;dr It might be interesting to consider a system where the "plotting live streaming data" parts of callbacks are separate from their "interpret documents and do something with them".